RFC-0011-coordinator-throttling.md

```diff
@@ -0,0 +1,134 @@
+
+While this approach is generally effective for distributing work, it may not always result in the most optimal allocation of resources.
+
+**Intermediate task scheduling**
+
+For intermediate task scheduling in Presto, the selection of workers is based on a hash partition count. Specifically, intermediate tasks are assigned to worker nodes according to the hash of the partition, which determines the distribution across available intermediate workers. The relevant code implementation can be found [here](https://github.com/prestodb/presto/blob/master/presto-main-base/src/main/java/com/facebook/presto/sql/planner/SystemPartitioningHandle.java#L164).
+This strategy does not consider the current system load when assigning intermediate tasks. 
+
+#### Resource Group throttling
+There are certain metrics like softMemoryLimit etc which exist at RG level to queue the workload, but it does not consider the actual utilization and is out of scope for this discussion.
+
+
+While we could improve task scheduling (pro-active measure to not let worker overload) through load balancing (more on this in alternative approaches), this alone would be insufficient given the nature of our overload scenario where all workers are saturated in short period of time which is due to admitting lot of heavy queries in short duration of time. 
+
+For immediate relief, we propose implementing a holistic admission control mechanism that provides reactive throttling of the overall workload by not admitting queries.
+
+### Goals
+- Decrease duration of overload on the cluster
+- Also decrease overall variance of execution time of queries (queuing is fine)
+
+## Proposed Implementation
+### Worker side changes
+- Added endpoint to gather worker load (`/v1/info/nodestats`)
+  - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).  
+- PR: https://github.com/prestodb/presto/pull/25686
+
+### Coordinator side changes
+#### Collecting worker load data
+  - Update DiscoveryManager to invoke this end point rather than the `/v1/info/state`
+  - Expose `getNodeLoadMetrics` on InternalNodeManager
+  - PR: https://github.com/prestodb/presto/pull/25688 
+#### Scheduling policies
+  - Implement node overload policies that can govern the policies to determine cluster overload
+  - Based on these policies and load collected from worker, improve the admission control logic globally to queue the query (irrespective of Resource Group)
+  - PR: https://github.com/prestodb/presto/pull/25689 
+
+
+## Metrics
+
+- Cluster overload
```

Comment:
Is this a binary metric?  Can you elaborate?

Reply:
Added details on the JMX metrics added.